### PR TITLE
v0.1 release

### PR DESCRIPTION
**Pull request by AntoineRondelet**

### CODE REVIEW

File: libff/common/utils.hpp [hunk 1]
```diff
@@ -31,16 +33,16 @@ size_t to_twos_complement(int i, size_t w);
...
+bit_vector int_list_to_bits(
+    const std::initializer_list<unsigned long> &l, const size_t wordsize);
 long long div_ceil(long long x, long long y);
 
 bool is_little_endian();
 
-std::string FORMAT(const std::string &prefix, const char* format, ...);
+std::string FORMAT(const std::string &prefix, const char *format, ...);
 
 /* A variadic template to suppress unused argument warnings */
```
Comment (AntoineRondelet) [inline libff/common/utils.hpp:44]: Use `///` comment

File: libff/algebra/curves/public_params.hpp [hunk 2]
```diff
@@ -9,95 +9,84 @@
 #define PUBLIC_PARAMS_HPP_
 #include <vector>
 
-namespace libff {
+namespace libff
+{
 
 /*
```
Comment (AntoineRondelet) [inline libff/algebra/curves/public_params.hpp:15]: Use `///`

File: libff/algebra/fields/fp.hpp [hunk 3]
```diff
@@ -13,172 +14,204 @@
...
+    static const constexpr bigint<n> &mod = modulus;
 #ifdef PROFILE_OP_COUNTS
     static long long add_cnt;
     static long long sub_cnt;
     static long long mul_cnt;
     static long long sqr_cnt;
     static long long inv_cnt;
 #endif
+
+    // The "base"/"ground" field
```
Comment (AntoineRondelet) [inline libff/algebra/fields/fp.hpp:57]: `///` (same applies for the lines below)

File: libff/algebra/fields/fp.hpp [hunk 4]
```diff
@@ -13,172 +14,204 @@
...
     static Fp_model<n, modulus> random_element();
-    static Fp_model<n, modulus> geometric_generator(); // generator^k, for k = 1 to m, domain size m
-    static Fp_model<n, modulus> arithmetic_generator();// generator++, for k = 1 to m, domain size m
 
-    friend std::ostream& operator<< <n,modulus>(std::ostream &out, const Fp_model<n, modulus> &p);
-    friend std::istream& operator>> <n,modulus>(std::istream &in, Fp_model<n, modulus> &p);
+    // generator^k, for k = 1 to m, domain size m
+    static Fp_model<n, modulus> geometric_generator();
+
+    // generator++, for k = 1 to m, domain size m
```
Comment (AntoineRondelet) [inline libff/algebra/fields/fp.hpp:147]: `///`

File: libff/algebra/fields/fp.tcc [hunk 5]
```diff
@@ -422,89 +464,80 @@ Fp_model<n,modulus>& Fp_model<n,modulus>::operator+=(const Fp_model<n,modulus>&
...
+            SUB_NEXTADD(24)      //
+            "done%=:\n\t"        //
+            :
+            : [A] "r"(this->mont_repr.data),
+              [B] "r"(other.mont_repr.data),
+              [mod] "r"(modulus.data)
+            : "cc", "memory", "%rax");
+    } else if (n == 5) {
+        __asm__(                 // Preserve alignment
+            SUB_FIRSTSUB()       //
```
Comment (AntoineRondelet) [inline libff/algebra/fields/fp.tcc:509]: What was the issue with the missing parenthesis?

File: libff/algebra/fields/fp12_2over3over2.hpp [hunk 6]
```diff
@@ -9,109 +9,148 @@
...
     static Fp2_model<n, modulus> non_residue;
-    static Fp2_model<n, modulus> Frobenius_coeffs_c1[12]; // non_residue^((modulus^i-1)/6) for i=0,...,11
-
-    my_Fp6 c0, c1;
-    Fp12_2over3over2_model() {};
-    Fp12_2over3over2_model(const my_Fp6& c0, const my_Fp6& c1) : c0(c0), c1(c1) {};
-
-    void clear() { c0.clear(); c1.clear(); }
-    void print() const { printf("c0/c1:\n"); c0.print(); c1.print(); }
+    // non_residue^((modulus^i-1)/6) for i=0,...,11
```
Comment (AntoineRondelet) [inline libff/algebra/fields/fp12_2over3over2.hpp:44]: Move up and use `///`

File: libff/algebra/fields/fp2.hpp [hunk 7]
```diff
@@ -9,113 +9,153 @@
...
-    my_Fp c0, c1;
-    Fp2_model() {};
-    Fp2_model(const my_Fp& c0, const my_Fp& c1) : c0(c0), c1(c1) {};
-
-    void clear() { c0.clear(); c1.clear(); }
-    void print() const { printf("c0/c1:\n"); c0.print(); c1.print(); }
-
-    static Fp2_model<n, modulus> zero();
-    static Fp2_model<n, modulus> one();
+    // Exposing the field extension degree via a static member
```
Comment (AntoineRondelet) [inline libff/algebra/fields/fp2.hpp:37]: `///`

File: libff/algebra/fields/fp4.hpp [hunk 8]
```diff
@@ -19,39 +20,63 @@
...
-    static my_Fp Frobenius_coeffs_c1[4]; // non_residue^((modulus^i-1)/4) for i=0,1,2,3
-
-    my_Fp2 c0, c1;
-    Fp4_model() {};
-    Fp4_model(const my_Fp2& c0, const my_Fp2& c1) : c0(c0), c1(c1) {};
+    // The field extension here is 4, but this is built as 2 over 2. As such,
+    // the value of `extension_degree` only contains the field extension used to
+    // obtain the top most level in the field tower, which is 2 here, not 4.
+    //
+    // TODO: Maybe chosen another name for this static member to avoid confusion
```
Comment (AntoineRondelet) [inline libff/algebra/fields/fp4.hpp:45]: choose

File: libff/algebra/fields/fp4.hpp [hunk 9]
```diff
@@ -19,39 +20,63 @@
...
+    // with the actual "total" extension degree.
+    // Note: `final_extension_degree` could be a good name.
+    static const size_t tower_extension_degree = 2;
 
-    void print() const { printf("c0/c1:\n"); c0.print(); c1.print(); }
-    void clear() { c0.clear(); c1.clear(); }
+    static my_Fp non_residue;
+    // non_residue^((modulus^i-1)/4) for i=0,1,2,3
```
Comment (AntoineRondelet) [inline libff/algebra/fields/fp4.hpp:51]: /// and move up

File: libff/algebra/fields/fp6_2over3.hpp [hunk 10]
```diff
@@ -13,54 +13,72 @@
...
     static my_Fp non_residue;
-    static my_Fp Frobenius_coeffs_c1[6]; // non_residue^((modulus^i-1)/6)   for i=0,1,2,3,4,5
-
-    my_Fp3 c0, c1;
-    Fp6_2over3_model() {};
-    Fp6_2over3_model(const my_Fp3& c0, const my_Fp3& c1) : c0(c0), c1(c1) {};
-
-    void print() const { printf("c0/c1:\n"); c0.print(); c1.print(); }
-    void clear() { c0.clear(); c1.clear(); }
+    // non_residue^((modulus^i-1)/6)   for i=0,1,2,3,4,5
```
Comment (AntoineRondelet) [inline libff/algebra/fields/fp6_2over3.hpp:43]: Move up and `///`

File: libff/algebra/scalar_multiplication/multiexp.tcc [hunk 11]
```diff
@@ -16,529 +16,788 @@
...
- */
-
-template<typename T, typename FieldT, multi_exp_method Method,
-    typename std::enable_if<(Method == multi_exp_method_naive), int>::type = 0>
-T multi_exp_inner(
-    typename std::vector<T>::const_iterator vec_start,
-    typename std::vector<T>::const_iterator vec_end,
-    typename std::vector<FieldT>::const_iterator scalar_start,
-    typename std::vector<FieldT>::const_iterator scalar_end)
+// Class holding the specialized multi exp implementations. Must implement a
```
Comment (AntoineRondelet) [inline libff/algebra/scalar_multiplication/multiexp.tcc:198]: ///

File: libff/algebra/scalar_multiplication/wnaf.hpp [hunk 12]
```diff
@@ -12,51 +13,57 @@
...
 #include <libff/algebra/fields/bigint.hpp>
+#include <vector>
 
-namespace libff {
+namespace libff
+{
 
 /**
  * Find the wNAF representation of the given scalar relative to the given
  * window size, reusing the given vector to store it.
```
Comment (AntoineRondelet) [inline libff/algebra/scalar_multiplication/wnaf.hpp:24]: Use `///`

File: libff/common/rng.tcc [hunk 13]
```diff
@@ -15,49 +15,48 @@
...
+namespace libff
+{
 
-template<typename FieldT>
-FieldT SHA512_rng(const uint64_t idx)
+template<typename FieldT> FieldT SHA512_rng(const uint64_t idx)
 {
-    assert(GMP_NUMB_BITS == 64); // current Python code cannot handle larger values, so testing here for some assumptions.
+    assert(
+        GMP_NUMB_BITS == 64); // current Python code cannot handle larger
```
Comment (AntoineRondelet) [inline libff/common/rng.tcc:29]: Move up

**Comment by AntoineRondelet** [comment]
This PR is still WIP as #61 targets it. If #61 gets merged, I'll remove the WIP from this PR and it's be ready to be reviewed.

**Review by dtebbs — APPROVED** [review]
LGTM

File: libff/algebra/fields/fp.tcc (hunk 5 above)
Comment (dtebbs) [reply libff/algebra/fields/fp.tcc:509]: Without the parentheses, I think clang-format would rearrange everything to make it unreadable.  IIRC, adding parentheses also made the macro consistent with several other macros.

Comment (AntoineRondelet) [reply libff/algebra/fields/fp.tcc:509]: I see, thanks

**Pull request merged**